Add new hook `actionProductGetAttributesGroupsAfter`

gwojcik7 opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.2.x
| Description?      | Added actionProductGetAttributesGroupsAfter hook which allows to modify product attributes groups after they are retrieved from the database.
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | Install the hook in the module and remove one of the options in the selected attribute group. Then check on the product card if indeed this option has disappeared from the combination selection.
| UI Tests          | Please run UI tests and paste here the link to the run. As we support MySQL and MariaDB in our tests, you have to launch 2 campaigns (by choosing both). [Visit the UI Tests repo and follow instructions](https://github.com/PrestaShop/ga.tests.ui.pr/).
| Fixed issue or discussion?     | 
| Related PRs       | 
| Sponsor company   | Arkonsoft

Hook actionProductGetAttributesGroupsAfter can allow, for example, to hide some options in combinations in terms of more complex logic. Combinations are used by stores in a variety of ways, not always exactly as intended. There may be a situation where a certain variant should not be sold in a certain currency, language version, country or paired with another product. The possibilities are many, and this hook will allow you to cut out unwanted options.

gwojcik7 committed:
add actionProductGetAttributesGroupsAfter hook

gwojcik7 committed:
add all method parameters to hook

ps-jarvis commented:
<!-- PR_TABLE_DESCRIPTION_ERROR -->
Hi, thanks for this contribution!

Please consider opening an issue before submitting a Pull Request:
* If it's a bug fix, it helps maintainers verify that the bug is effectively due to a defect in the code, and that it hasn't been fixed already.
* It can help trigger a discussion about the best implementation path before a single line of code is written.
* It may lead the Core Product team to mark that issue as a priority, further attracting the maintainers' attention.

(Note: this is an automated message, but answering it will reach a real human)

kpodemski reviewed (commented):
Hello @gwojcik7 !

Thanks for contributing to PrestaShop 💪🏻 

At first glance, it all makes sense, and I understand the need. :D

I've been thinking about scenarios like this for some time already, and I wondered, " Wouldn't it be better to come up with a solution that lets us modify the query directly?"

Why?

- it's cleaner
- It could help avoid n+1 if we need to retrieve more data

If we rework the query to be based on `DbQuery` class, then I think it could be possible just to pass the `DbQuery` object and let it modify it by the hook

To be clear, I don't want to block the PR or ask you to do that. I'm opening a discussion on how to approach such scenarios better because I can see potential in places where we retrieve the products, etc.

Any thoughts on this?

also, Mr @Hlavtox what do you think? :)

jolelievre commented:
The proposition seems ok, but the questions from @kpodemski are completely relevant, maybe there are alternatives to implement this better. In case we don't find a better solution this one seems ok nonetheless

gwojcik7 commented:
@kpodemski You are right, doing it at the SQL level is more efficient. I will prepare this change.

gwojcik7 committed:
migrate product attribute groups SQL to DbQuery class

gwojcik7 committed:
add actionProductGetAttributesGroupsBefore to manipulate query class

gwojcik7 commented:
@kpodemski @jolelievre I rewrote the SQL query to use DbQuery and added the `actionProductGetAttributesGroupsBefore` hook for query manipulation. However, I also kept the `actionProductGetAttributesGroupsAfter` hook, as I see some use cases where it might be necessary to operate on the actual data retrieved from the database.

boherm reviewed (approved):
Nice!

paulnoelcholot reviewed (commented):
Automated test are green : https://github.com/paulnoelcholot/testing_pr/actions/runs/14448226928

gericfo commented:
Hello @gwojcik7 

Please note that the “Needs autoupgrade PR” label has been automatically added to your Pull Request. This means that the changes made in your PR must also be reported in the Autoupgrade module, responsible for running the update process of PrestaShop stores.

When there's a modification to the db-structure.sql file or if a Doctrine entity has been modified, added, or deleted, it impacts the database structure. Modifying the list of hooks and feature flags has consequences on the update process as well. These changes can affect how the application interacts with the database and how update procedures are handled. It's important to consider these aspects when contributing to ensure a smooth update experience for users.

That's why this contribution should be backported, with another corresponding PR, on the Autoupgrade module's GitHub repository which is available here: [Autoupgrade Repository](https://github.com/PrestaShop/autoupgrade)
Contribution to the Autoupgrade module documentation: [Documentation Link](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/upgrade-module/)

Don't forget to add a comment in this current thread with the link to this PR once it has been created.

If you are sure that your Pull Request doesn’t require additional actions, ping [committers](https://www.prestashop-project.org/project-organization/people-and-roles/#committers) and it will be checked.

Thank you.
